dali/kernels/signal/window/extract_windows_gpu.cuh

```diff
@@ -0,0 +1,301 @@
+ * CUDA block size is fixed kBlock x kBlock - it processes kBlock values from kBlock windows.
+ * A logical block may contain multiple CUDA blocks - the multiplier is always
+ * equal to 1 << block_shift.
+ */
+template <typename Dst, typename Src>
+__global__ void ExtractWindowsBatchedKernel(
+    const SampleDesc *samples,
+    const window::BlockDesc *blocks,
+    int windows_per_block,
+    const float *window,
+    int win_len,
+    int win_center,
+    int step,
+    bool reflect) {
+  int block_idx = blockIdx.x;
+  BlockDesc blk = blocks[block_idx];
+  SampleDesc sample = samples[blk.sample_idx];
+  const Src *__restrict__ src = static_cast<const Src*>(sample.input);
+  Dst *__restrict__ dst = static_cast<Dst*>(sample.output);
+  ptrdiff_t stride = sample.output_stride;
+  int num_windows = sample.num_windows;
+  ptrdiff_t length = sample.length;
+
+  int block_end = min(blk.start + windows_per_block, num_windows);
+
+  // NOTE: This loop is dynamically uniform. Introducing any thread-index dependent condition
+  // will break the code, leading to undefined behavior (including hangs).
+  for (int pos = blk.start, page = 0; pos < block_end; pos += kBlock, page = 1-page) {
+    ExtractWindowsBlock<2>(
+      pos,                        // first window index
+      dst, num_windows, stride,   // output
+      src, length,                // input
+      window, win_len, win_center, step, reflect,  // windowing options
+      page);  // page-flipped temporary buffer avoids additional __syncthreads
+  }
+}
+
+}  // namespace window
+
+template <typename Dst, typename Src>
+struct ExtractWindowsGPUImpl {
+  using SampleDesc = window::SampleDesc;
+  using BlockDesc = window::BlockDesc;
+
+  KernelRequirements Setup(
+      KernelContext &context,
+      const InListGPU<Src, 1> &input,
+      const ExtractWindowsArgs &args,
+      bool concatenate,
+      int out_win_length = -1) {
+    block_dim  = dim3(window::kBlock, window::kBlock);
+
+    const auto &lengths = input.shape;
+
+    this->args = args;
+    this->concatenate = concatenate;
+    if (out_win_length < 0)
+      out_win_length = this->args.window_length;
+    else if (out_win_length < this->args.window_length)
+      this->args.window_length = out_win_length;
+
+    int N = lengths.num_samples();
+    int ygrid = div_ceil(this->args.window_length, window::kBlock);
+
+    int64_t total_windows = 0;
+
+    TensorListShape<2> out_shape;
+    out_shape.resize(concatenate ? 1 : N);
+
+    windows_per_block = window::kBlock;
+    int xgrid = 0;
+    for (int i = 0; i < N; i++) {
+      int nwin = args.num_windows(lengths[i][0]);
+      total_windows += nwin;
+      int blocks = div_ceil(nwin, windows_per_block);
+      xgrid += blocks;
+      if (!concatenate) {
+        out_shape.set_tensor_shape(i, { out_win_length, nwin });
+      }
+    }
+    if (concatenate) {
+      out_shape.set_tensor_shape(0, { out_win_length, total_windows });
+    }
+
+    while (xgrid > 0x10000 && xgrid > N) {
```

Comment:
please make 0x10000 a named constant

Reply:
I can, I just don't know if it really helps to have `kMaxBlocks = 0x10000` defined in previous line and used just once...